### PR TITLE
sqlite3: add 3.11 additions

### PR DESCRIPTION
**Pull request by JelleZijlstra**
- Blob from python/cpython#30680 (and anticipating that python/cpython#91550 will be merged)
- Aggregate window functions from python/cpython#20903
- Serialize/deserialize from python/cpython#26728
- Limit setting from python/cpython#28463

### CODE REVIEW

File: stdlib/sqlite3/dbapi2.pyi [hunk 1]
```diff
@@ -128,6 +129,18 @@ class _AggregateProtocol(Protocol):
     def step(self, value: int) -> None: ...
     def finalize(self) -> int: ...
 
+class _SingleParamWindowAggregateClass(Protocol):
+    def step(self, __param: Any) -> None: ...
+    def inverse(self, __param: Any) -> None: ...
```
Comment (srittau) [inline stdlib/sqlite3/dbapi2.pyi:134]: As an implementation could return a value, but users should ignore it:

```suggestion
    def step(self, __param: Any) -> object: ...
    def inverse(self, __param: Any) -> object: ...
```

File: stdlib/sqlite3/dbapi2.pyi [hunk 2]
```diff
@@ -146,8 +159,23 @@ class Connection:
...
     def __init__(self, *args: Any, **kwargs: Any) -> None: ...
     def close(self) -> None: ...
+    if sys.version_info >= (3, 11):
+        def blobopen(self, __table: str, __column: str, __row: int, *, readonly: bool = ..., name: str = ...) -> Blob: ...
+
     def commit(self) -> None: ...
     def create_aggregate(self, name: str, n_arg: int, aggregate_class: Callable[[], _AggregateProtocol]) -> None: ...
+    if sys.version_info >= (3, 11):
+        # num_params determines how many params will be passed to the aggregate class. We provide an overload
+        # for the case where num_params = 1, which is expected to be the common case.
```
Comment (srittau) [inline stdlib/sqlite3/dbapi2.pyi:169]: We could also special case -1 with a protocol:

```
class _AnyParamWindowAggregateClass(Protocol):
    def step(self, *params: Any) -> object: ...
    def inverse(self, *params: Any) -> object: ...
    def value(self) -> _SqliteData: ...
    def finalize(self) -> _SqliteData: ...
```

File: stdlib/sqlite3/dbapi2.pyi [hunk 3]
```diff
@@ -146,8 +159,23 @@ class Connection:
...
+        @overload
+        def create_window_function(
+            self, __name: str, __num_params: Literal[1], aggregate_class: Callable[[], _SingleParamWindowAggregateClass]
+        ) -> None: ...
+        @overload
+        def create_window_function(
+            self, __name: str, __num_params: int, aggregate_class: Callable[[], _WindowAggretateClass]
+        ) -> None: ...
```
Comment (srittau) [inline stdlib/sqlite3/dbapi2.pyi:177]: According to the documentation, `aggregate_class` can also be `None` to clear a window function:

```
@overload
def create_window_function(self, __name: str, __num_params: int, aggregate_class: None) -> None: ...
```

**Review by srittau — CHANGES_REQUESTED** [review]
Exciting changes to sqlite!

File: stdlib/sqlite3/dbapi2.pyi (hunk 1 above)
Comment (srittau) [reply stdlib/sqlite3/dbapi2.pyi:134]: The same for `_WindowAggretateClass`.

File: stdlib/sqlite3/dbapi2.pyi (hunk 3 above)
Comment (srittau) [reply stdlib/sqlite3/dbapi2.pyi:177]: Also, `__aggregate_class` in all three overloads.

Comment (JelleZijlstra) [reply stdlib/sqlite3/dbapi2.pyi:188]: Good point, done

File: stdlib/sqlite3/dbapi2.pyi (hunk 1 above)
Comment (JelleZijlstra) [reply stdlib/sqlite3/dbapi2.pyi]: Done

File: stdlib/sqlite3/dbapi2.pyi [hunk 4]
```diff
@@ -125,9 +126,27 @@ if sys.version_info < (3, 8):
...
+    def step(self, value: int) -> object: ...
     def finalize(self) -> int: ...
 
+class _SingleParamWindowAggregateClass(Protocol):
+    def step(self, __param: Any) -> object: ...
+    def inverse(self, __param: Any) -> object: ...
+    def value(self) -> _SqliteData: ...
+    def finalize(self) -> _SqliteData: ...
+
+class _AnyParamWindowAggreateClass(Protocol):
```
Comment (AlexWaygood) [inline stdlib/sqlite3/dbapi2.pyi:138]:
```suggestion
class _AnyParamWindowAggregateClass(Protocol):
```

File: stdlib/sqlite3/dbapi2.pyi [hunk 5]
```diff
@@ -125,9 +126,27 @@ if sys.version_info < (3, 8):
...
+    def step(self, *args: Any) -> object: ...
+    def inverse(self, *args: Any) -> object: ...
+    def value(self) -> _SqliteData: ...
+    def finalize(self) -> _SqliteData: ...
+
+class _WindowAggretateClass(Protocol):
```
Comment (AlexWaygood) [inline stdlib/sqlite3/dbapi2.pyi:144]:
```suggestion
class _WindowAggregateClass(Protocol):
```

File: stdlib/sqlite3/dbapi2.pyi [hunk 6]
```diff
@@ -146,8 +165,28 @@ class Connection:
...
+        # num_params determines how many params will be passed to the aggregate class. We provide an overload
+        # for the case where num_params = 1, which is expected to be the common case.
+        @overload
+        def create_window_function(
+            self, __name: str, __num_params: Literal[1], __aggregate_class: Callable[[], _SingleParamWindowAggregateClass] | None
+        ) -> None: ...
+        # And for num_params = -1, which means the aggregate must accept any number of parameters.
+        @overload
+        def create_window_function(
+            self, __name: str, __num_params: Literal[-1], __aggregate_class: Callable[[], _AnyParamWindowAggreateClass] | None
```
Comment (AlexWaygood) [inline stdlib/sqlite3/dbapi2.pyi:183]:
```suggestion
            self, __name: str, __num_params: Literal[-1], __aggregate_class: Callable[[], _AnyParamWindowAggregateClass] | None
```

File: stdlib/sqlite3/dbapi2.pyi [hunk 7]
```diff
@@ -146,8 +165,28 @@ class Connection:
...
+        ) -> None: ...
+        @overload
+        def create_window_function(
+            self, __name: str, __num_params: int, __aggregate_class: Callable[[], _WindowAggretateClass] | None
```
Comment (AlexWaygood) [inline stdlib/sqlite3/dbapi2.pyi:187]:
```suggestion
            self, __name: str, __num_params: int, __aggregate_class: Callable[[], _WindowAggregateClass] | None
```

**Review by AlexWaygood — CHANGES_REQUESTED** [review]
A few typos :)

File: stdlib/sqlite3/dbapi2.pyi [hunk 8]
```diff
@@ -253,3 +297,15 @@ if sys.version_info < (3, 8):
...
+    class Blob:
+        def close(self) -> None: ...
+        def read(self, __length: int = ...) -> bytes: ...
+        def write(self, __data: bytes) -> None: ...
+        def tell(self) -> int: ...
+        # whence must be one of os.SEEK_SET, os.SEEK_CUR, os.SEEK_END
+        def seek(self, __offset: int, __whence: int = ...) -> None: ...
+        def __len__(self) -> int: ...
+        def __enter__(self: Self) -> Self: ...
+        def __exit__(self, __typ: object, __val: object, __tb: object) -> Literal[False]: ...
```
Comment (AlexWaygood) [inline stdlib/sqlite3/dbapi2.pyi:311]: I think this is going to anger my unreleased Y036 check over at flake8-pyi :/ https://github.com/PyCQA/flake8-pyi/commit/b7f7e9a7cfab24862ae8bf0f620ad8b84e740d96

**Review by srittau — APPROVED** [review]
LGTM, apart from the things @AlexWaygood pointed out.

File: stdlib/sqlite3/dbapi2.pyi (hunk 8 above)
Comment (JelleZijlstra) [reply stdlib/sqlite3/dbapi2.pyi:311]: hm, do you think we should change it? This `__exit__` method really does take `object`; that's what the AC annotations say.

Comment (AlexWaygood) [reply stdlib/sqlite3/dbapi2.pyi:311]: 'Kay, I'll work up a patch for flake8-pyi.

File: stdlib/sqlite3/dbapi2.pyi (hunk 4 above)
Comment (JelleZijlstra) [reply stdlib/sqlite3/dbapi2.pyi]: Thanks, applied all these manually

**Review by AlexWaygood — APPROVED** [review]
Typos have been corrected to my satisfaction :)

**Pull request merged**